applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_displacement_response_function.cpp

```diff
@@ -65,20 +68,24 @@ namespace Kratos
 
         rResponseGradient.clear();
 
-        if( rAdjointElement.Id() == mpNeighboringElement->Id() )
-        {
-            DofsVectorType dofs_of_element;
-            mpNeighboringElement->GetDofList(dofs_of_element, rProcessInfo);
-
-            const DoubleVariableType& r_traced_adjoint_dof =
-                KratosComponents<DoubleVariableType>::Get(std::string("ADJOINT_") + mTracedDofLabel);
-
-            for(IndexType i = 0; i < dofs_of_element.size(); ++i)
-            {
-                if (dofs_of_element[i]->Id() == mpTracedNode->Id() &&
-                    dofs_of_element[i]->GetVariable() == r_traced_adjoint_dof)
-                {
-                    rResponseGradient[i] = -1;
+        std::map<std::pair<IndexType, IndexType>, IndexType>::iterator it;
+        for (it = mElementNodeMap.begin(); it != mElementNodeMap.end(); ++it) {
+            const auto element_id = (it->first).first;
+            const auto node_id = it->second;
+
+            if( rAdjointElement.Id() == element_id ) {
+                const Variable<double>* adjoint_solution_variable = &KratosComponents<Variable<double>>::Get("ADJOINT_" + mTracedDofLabel + "_X");
+                Element::Pointer p_neighboring_element = mrModelPart.pGetElement(element_id);
+                DofsVectorType dofs_of_element;
+                p_neighboring_element->GetDofList(dofs_of_element, rProcessInfo);
+
+                for(IndexType i = 0; i < dofs_of_element.size(); ++i) {
+                    if (dofs_of_element[i]->Id() == node_id &&
+                        dofs_of_element[i]->GetVariable() == *adjoint_solution_variable) {
+                        rResponseGradient[i]   = -1 * mResponseDirection[0];
```

Comment:
I think like this you are assembling the contribution for one node multiple times if it is part of multiple elements. 

EDIT: It seems to be correct, because you are also checking for the node id - i am not convinced yet that this logic can't be improved.

Reply:
in any case, once you entered this if, you can break the loop.

Reply:
correct this should not happen because one element is linked with one node by the map